v5/core/parameterized_url.go

```diff
@@ -0,0 +1,83 @@
+package core
+
+// (C) Copyright IBM Corp. 2021.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+//
+// constructServiceURL returns a service URL that is constructed by formatting a parameterized URL.
+//
+// Parameters:
+//
+// parameterizedUrl: URL that contains variable placeholders, e.g. "{scheme}://ibm.com".
+//
+// defaultUrlVariables: map from variable names to default values.
+//		Each variable in the parameterized URL must have a default value specified in this map.
+//
+// providedUrlVariables: map from variable names to desired values.
+// 		If a variable is not provided in this map,
+// 		the default variable value will be used instead.
+//
+func constructServiceURL(
+	parameterizedUrl string,
+	defaultUrlVariables map[string]string,
+	providedUrlVariables map[string]string,
+) (string, error) {
+
+	// If null was passed, we set the variables to an empty map.
+	// This results in all default variable values being used.
+	if providedUrlVariables == nil {
+		providedUrlVariables = map[string]string{}
+	}
+
+	// Verify the provided variable names.
+	for providedName := range providedUrlVariables {
+		if _, ok := defaultUrlVariables[providedName]; !ok {
+			// Get all variable names (the keys of the default variables map).
+			acceptedNames := make([]string, len(defaultUrlVariables))
+
+			i := 0
+			for name := range defaultUrlVariables {
+				acceptedNames[i] = name
```

Comment:
```suggestion
				acceptedNames = append(acceptedNames, name)
```
When adding items to a slice in go, I think it is more idiomatic to use "append".  That way, you don't need to worry about the index variable, etc.
There is a slight performance hit in using append, but in this case it shouldn't be a concern since we're only dealing with a small number of variable names.

Reply:
I've changed the code to use `append` now.

I don't think I can combine the two loops because the first is iterating over the provided variables, and the second is iterating over the default variables. The first loop helps the user by catching if they misspelled a variable name. If we just looped over the default variables, the default value would be used instead of the intended provided value, so the user could get a URL that they didn't want.

I kept the creation of the `acceptedNames` inside the loop so it's only invoked if a provided variable name is invalid. Otherwise, we don't need to compute it.

Reply:
I see what you're saying about not being able to combine the two loops.   I didn't realize before that you were looping over entries in the two different maps :)